examples/weakref.cpp

```diff
@@ -0,0 +1,279 @@
+#include <stdio.h>
+
+#include <jsapi.h>
+#include <jsfriendapi.h>
+#include <js/CompilationAndEvaluation.h>
+#include <js/Initialization.h>
+#include <js/Promise.h>
+#include <js/Realm.h>
+#include <js/SourceText.h>
+#include <mozilla/Unused.h>
+
+#include "boilerplate.h"
+
+// This example illustrates what you have to do in your embedding to make
+// WeakRef and FinalizationRegistry work. Without notifying SpiderMonkey when to
+// clear out WeakRefs and run FinalizationRegistry callbacks, they will appear
+// not to work correctly.
+//
+// See 'boilerplate.cpp' for the parts of this example that are reused in many
+// simple embedding examples.
+
+// This function silently ignores errors in a way that production code probably
+// wouldn't.
+static void LogPendingException(JSContext* cx) {
+  // Nothing we can do about uncatchable exceptions.
+  if (!JS_IsExceptionPending(cx)) return;
+
+  JS::ExceptionStack exnStack{cx};
+  if (!JS::StealPendingExceptionStack(cx, &exnStack)) return;
+
+  JS::ErrorReportBuilder builder{cx};
+  if (!builder.init(cx, exnStack, JS::ErrorReportBuilder::NoSideEffects)) {
+    return;
+  }
+  JS::PrintError(stderr, builder, /* reportWarnings = */ false);
+}
+
+// FIXME: Is it really necessary to have a custom job queue in order to use
+// WeakRef and FinalizationRegistry? (And if not, is it a good idea to have one
+// anyway in this example because that's how most embeddings will use it?)
```

Comment:
This is one thing I'm not really sure about. I guess we could just clear kept objects and run finalization registry cleanups in the `gc()` function and the example would still work?

Reply:
Given the FinalizationRegistry is [defined based on job](https://tc39.es/ecma262/#sec-host-cleanup-finalization-registry), I think it's better having some kind of job queue, but yes, it doesn't have to be `JS::JobQueue`-based one, and just "running the cleanup functions at some time future" is only the requirement here.

Also, for the purpose of example, the regular promise job handling can be omitted (or maybe add some comment that `m_queue` handling is optional?)